File: java/main/src/main/java/TwoDmArrays.java
public class TwoDmArrays {
    public static void main(String[] args) {
        int[][] bombs1 = {{0, 2}, {2, 0}};
        mineSweeper(bombs1, 3, 3);
        // mineSweeper(bombs1, 3, 3) should return:
        // [[0, 1, -1],
        //  [1, 21, 1],
        //  [-1, 1, 0]]

        int[][] bombs2 = {{0, 0}, {0, 1}, {1, 2}};
        mineSweeper(bombs2, 3, 4);
        // mineSweeper(bombs2, 3, 4) should return:
        // [[-1, -1, 2, 1],
        //  [2, 3, -1, 1],
        //  [0, 1, 1, 1]]

        int[][] bombs3 = {{1, 1}, {1, 2}, {2, 2}, {4, 3}};
        mineSweeper(bombs3, 5, 5);
        // mineSweeper(bombs3, 5, 5) should return:
        // [[1, 2, 2, 1, 0],
        //  [1, -1, -1, 2, 0],
        //  [1, 3, -1, 2, 0],
        //  [0, 1, 2, 2, 1],
        //  [0, 0, 1, -1, 1]]
    }

    public static int[][] mineSweeper(int[][] a, int b, int c) {
        int[][] field = new int[b][c];
        for (int[] bomb : a) {
            int rowIndex = bomb[0];
            int colIndex = bomb[1];
            field[rowIndex][colIndex] = -1;
            for (int i = rowIndex - 1; i < rowIndex + 2; i++) {
                for (int j = colIndex - 1; j < colIndex + 2; j++) {
                    if (0 <= i && i < b && 0 <= j && j < c && field[i][j] != -1) {
                        field[i][j] += 1;
                    }
                }
            }
        }
        return field;
    }
}

File: javascript/twoDmArrays.js
var bombs1 = [
    [0, 2],
    [2, 0]
];
console.log(minesweeper(bombs1, 3, 3));
// [[0, 1, -1],
// [1, 2, 1],
// [-1, 1, 0]]

var bombs2 = [
    [0, 0],
    [0, 1],
    [1, 2]
];
console.log(minesweeper(bombs2, 3, 4));
// [[-1, -1, 2, 1],
// [2, 3, -1, 1],
// [0, 1, 1, 1]]

var bombs3 = [
    [1, 1],
    [1, 2],
    [2, 2],
    [4, 3]
];
console.log(minesweeper(bombs3, 5, 5));
// [[1, 2, 2, 1, 0],
// [1, -1, -1, 2, 0],
// [1, 3, -1, 2, 0],
// [0, 1, 2, 2, 1],
// [0, 0, 1, -1, 1]]

function minesweeper(bombs, numRows, numCols) {
    var field = [];
    for (var i = 0; i < numRows; i++) {
        field[i] = [];
        for (var j = 0; j < numCols; j++) {
            field[i][j] = 0;
        }
    }
    for (const bomb of bombs) {
        const rowIndex = bomb[0];
        const colIndex = bomb[1];
        field[rowIndex][colIndex] = -1;
        for (let i = rowIndex - 1; i < rowIndex + 2; i++) {
            for (let j = colIndex - 1; j < colIndex + 2; j++) {
                if (0 <= i && i < numRows &&
                    0 <= j && j < numCols && field[i][j] != -1) {
                    field[i][j] += 1;
                }
            }
        }
    }
    return field;
}
